RFC/RFC-14-p2p-adapter.md

```diff
@@ -0,0 +1,348 @@
+    Peer5->>RS: connected to []
+```
+
+### 3. The routing service create a routing table for each peer
+
+A route is a list of peer ids. It doesn't contain the source or destination, so for example if a->b->c:
+- the route between a and c is [ b ]
+- and the route between b and c is []
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer3
+    participant Peer4
+    participant Peer5
+    participant RS as Routing Service
+
+    RS->>Peer1: { peer2: [], peer3: [], peer4: [] }
+    RS->>Peer2: { peer1: [], peer3: [], peer4: [ peer1 ] }
+    RS->>Peer3: { peer1: [], peer2: [], peer4: [ peer1 ] }
+    RS->>Peer4: { peer1: [], peer2: [ peer1 ], peer3: [ peer1 ] }
+    RS->>Peer5: {}
+```
+
+### Peer1 sends a message directly to peer2
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant MS as Messaging Service
+
+    Peer1->>Peer2: peer1 sends message directly to peer2
+```
+
+### Peer2 sends message to peer4 through peer1
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer4
+    participant MS as Messaging Service
+
+    Peer2->>Peer1: peer2 sends message to peer4 through peer1
+    Peer1->>Peer4: peer2 sends message to peer4 through peer1
+
+```
+
+### Peer1 sends message though messaging service to peer5
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer5
+    participant MS as Messaging Service
+
+      Peer1->>MS: peer1 sends message though messaging service to peer5
+      MS->>Peer5: peer1 sends message though messaging service to peer5
+```
+
+## Example: A connection is lost while relaying a package
+
+Let's assume the mesh is connected as: `peer1 <-> peer2 <-> peer3 <-> peer4`, and peer1 needs to send a message to peer4
+
+```mermaid
+sequenceDiagram
+    participant Peer1
+    participant Peer2
+    participant Peer3
+    participant Peer4
+    participant MS as Messaging Service
+    participant RS as Routing Service
+
+    Peer1->>Peer2: send message to relay to peer3
+
+    Note over Peer3: peer3 is disconnected
+
+    Peer2->>RS: update status: lost connection to peer3
+    Peer2->>MS: relay this message to peer4
+    MS->>Peer4: relay this message to peer4
+```
+
+## Definitions
+
+```typescript
+
+type Address = string
+
+// A route is a list of peer ids. It doesn't contain the source or destination, so for example if a->b->c
+// the route between a and c is [ b ]
+// and the route between b and c is []
+type Route = Address[]
+
+// A map between peer and a route. If a peer is missing, it means there is no direct or indirect P2P connection to the peer, so the messaging server should be used
+type PeerRoutingTable = Map<Address, Route>
+```
+
+## Messaging service
+
+To establish a P2P webrtc connection, peers will exchange signals with each other using the messaging service. This service will act as a fallback when no P2P route is available to deliver a message.
```

Comment:
So... the messaging service has the responsibility of doing the signalling?

Reply:
(sorry I missed this comment): yes, it does, right now the topics service is doing that but we could leverage this new service, which is more precise